website/docs/r/ruleset.html.markdown

```diff
@@ -0,0 +1,160 @@
+same syntax used in custom Firewall Rules. Cloudflare uses the Ruleset Engine
+in different products, allowing you to configure several products using the same
+basic syntax.
+
+## Example Usage
+
+```hcl
+# Magic Transit
+resource "cloudflare_ruleset" "magic_transit_example" {
+  account_id  = "d41d8cd98f00b204e9800998ecf8427e"
+  name        = "account magic transit"
+  description = "example magic transit ruleset description"
+  kind        = "root"
+  phase       = "magic_transit"
+
+  rules {
+    action = "allow"
+    expression = "tcp.dstport in { 32768..65535 }"
+    description = "Allow TCP Ephemeral Ports"
+  }
+}
+
+# Zone level managed WAF
+resource "cloudflare_ruleset" "zone_level_managed_waf" {
+  zone_id     = "cb029e245cfdd66dc8d2e570d5dd3322"
+  name        = "managed WAF"
+  description = "managed WAF ruleset description"
+  kind        = "zone"
+  phase       = "http_request_firewall_managed"
+
+  rules {
+    action = "execute"
+    action_parameters {
+      id = "efb7b8c949ac4650a09736fc376e9aee"
+    }
+    expression = "true"
+    description = "Execute Cloudflare Managed Ruleset on my zone-level phase ruleset"
+    enabled = true
+  }
+}
+
+# Zone level WAF with category based overrides
+resource "cloudflare_ruleset" "zone_level_managed_waf_with_category_based_overrides" {
+  zone_id     = "cb029e245cfdd66dc8d2e570d5dd3322"
+  name        = "managed WAF with category based overrides"
+  description = "managed WAF with category based overrides ruleset description"
+  kind        = "zone"
+  phase       = "http_request_firewall_managed"
+
+  rules {
+    action = "execute"
+    action_parameters {
+      id = "efb7b8c949ac4650a09736fc376e9aee"
+      overrides {
+        categories {
+          category = "wordpress"
+          action = "block"
+          enabled = true
+        }
+
+        categories {
+          category = "joomla"
+          action = "block"
+          enabled = true
+        }
+      }
+    }
+
+    expression = "true"
+    description = "overrides to only enable wordpress rules to block"
+    enabled = false
+  }
+}
+```
+
+## Argument Reference
+
+The following arguments are supported:
+
+* `account_id` - (Optional) The ID of the account where the Ruleset is being created. Conflicts with `"zone_id"`.
+* `description` - (Required) Brief summary of the Ruleset and the intended use.
+* `kind` - (Required) Type of Ruleset to create. Valid values are `"custom"`, `"managed"`, `"root"`, `"schema"` or `"zone"`.
+* `name` - (Required) Name for the Ruleset.
+* `phase` - (Required) Point in the request/response lifecycle where the Ruleset should be created. Valid values are `"ddos_l4"`, `"ddos_l7"`, `"http_request_firewall_custom"`, `"http_request_firewall_managed"`, `"http_request_late_transform"`, `"http_request_main"`, `"http_request_sanitize"`, `"http_request_transform"`, `"http_response_firewall_managed"`, `"magic_transit"` or `"http_ratelimit"`.
```

Comment:
we show all the phases here for two reasons. the first is that this resource covers all ruleset usage (not just WAF) and the second is that in the future, the valid values will be generated by the SDKs themselves using the `ValidateFunc` values which here, will show all of them. this keeps it consistent with our path forward and covering all uses. 

it will also provide a way to demonstrate deprecations from the SDK into the upstream systems like terraform.